### DISCUSSION TITLE
Using patterns as part of conditional object assignment

### DISCUSSION THREAD

**Discussion by WhitWaldo** [dotnet/csharplang#6652] [Language Ideas] [answered]
I instrument a lot of telemetry (with Application Insights) throughout my applications and I regularly have to populate dictionaries as part of this. I often know most of the keys I'm initializing the dictionary with, and do so as follows:

```csharp
_telemetryClient.TrackTrace("Doing a thing", SeverityLevel.Information, new Dictionary<string,string>{
  { "userId", userId.ToString() }
});
```

Sometimes I only optionally want to include some keys in the dictionary if they aren't null and contain a value. If this happens, I have to initialize my dictionary on a separate line and put an if statement around the logic to add additional keys to the dictionary:

```csharp
var telemetryParams = new Dictionary<string,string>{
  {"userId", userId.ToString()}
};
if (assignmentId != null) {
  telemetryParams.Add("assignmentId", assignmentId);
}
_telemetryClient.TrackTrace("Doing a thing", SeverityLevel.Information, telemetryParams);
```

Especially since I'm looking to simply instrument the code and move on, this now spans several additional lines just to formally and conditionally add additional entries to the dictionary after creating it. Since C# has seen the introduction of pattern matching in other ways, I'd like to request that object initialization be another avenue for them, so we get something like:

```csharp
_telemetryClient.TrackTrace("Doing a thing", SeverityLevel.Information, new Dictionary<string,string>{
  { "userId", userId.ToString() },
  { "assignmentId", assignmentId } when assignmentId is not null
});
```

And just like that I get significantly improved readability and a more concise syntax. 

Thank you for the consideration!

**Comment by svick** [comment]
Would creating a class that inherits from `Dictionary` and adds custom logic to its `Add` method be an option for you? I.e.:

```c#
class NotNullStringDictionary : Dictionary<string, string>
{
    public new void Add(string key, string? value)
    {
        if (value != null)
            base.Add(key, value);
    }
}

var telemetryParams = new NotNullStringDictionary{
    {"userId", userId.ToString()},
    {"assignmentId", assignmentId}
};
```

**Comment by jnm2** [comment, marked as answer]
The collection literals proposal (https://github.com/dotnet/csharplang/issues/5354) would currently allow this syntax:

```cs
_telemetryClient.TrackTrace("Doing a thing", SeverityLevel.Information, [
  "userId": userId.ToString(),
  ..assignmentId is not null ? ["assignmentId": assignmentId] : [],
]);
```

**Reply by WhitWaldo** [reply]
Your linked proposal is even better than what I'd considered. I'm all for that.
